src/aging_session/memory.clj

```diff
@@ -43,6 +64,67 @@
     (assoc session-map key entry)
     (dissoc session-map key)))
 
+(def ^:private internals-atom (atom {:enc-key nil
+                                     :mac-key nil
+                                     :counter 0}))
+
+(defn get-crypto-keys
+  []
+  (let [{enc-key :enc-key
+         mac-key :mac-key} @internals-atom]
+    (if (and enc-key mac-key)                               ; if keys are present - use them
+      [enc-key mac-key]
+      (do                                                   ; otherwise generate new keys and attempt to set them
+        (swap! internals-atom update-in [:enc-key] #(if-not % (nonce/random-bytes 32) %))
+        (swap! internals-atom update-in [:mac-key] #(if-not % (nonce/random-bytes 128) %))
+        (get-crypto-keys)))))                               ; retrieve the keys again, as something else may have set them in the meantime
+
+(defn decrypt
+  [iv key data]
+  (try
+    (nippy/thaw (crypto/decrypt-cbc (crypto/block-cipher :aes :cbc) data key iv))
+    (catch Exception e
+      (log/error "Error decrypting cookie: " e)
```

Comment:
what's the alternative? returning nil denies access - we treat wrong cookie as no cookie.

Reply:
What I'm thinking is that this is a "deal breaker" for the rest of the code, right? If this step fails, what's the point in doing anything else?

Catching and treating exceptions is something that makes more sense to be done when a step that is not mandatory fails or you can do something to recover from that - for instance, maybe get the value somewhere else? But this is neither the case here.

Will also be less obvious when debugging if things simply explode. 

Plus, you will be storing and accessing Redis more with this implementation, because you'll be reading and storing things to it when it's not necessary. The `nil` does not short-circuit that. And I don't recall how Redis billing is done on AWS, if its per use or not. But there are databases there - like DynamoDB - that bill you according to access. So this may be another reason for changing it.

Reply:
>If this step fails, what's the point in doing anything else?

we still have to process the response. an invalid cookie is just that - an invalid cookie. 

we shouldn't throw random 500s for invalid cookies - that's not user friendly and would only be confusing. 

processing an invalid cookie as no cookie makes more sense to me. also removes issues with users (if a non-malicious user somehow manages to have a corrupt cookie, throwing a 500 would not reset it for them. treating them as having no cookie and issuing a new one would)

as for redis - that's probably better to discuss in the private repo..........